File: 2023/python/42.py
with open("../test-cases/inp4.txt", "r") as inpfile:
    cards = inpfile.readlines()
    totalScratchCards = {}
    
    for e in range(1, len(cards)+1):
        totalScratchCards[e] = 1
        
    cNo = totalSC = 0
    for card in cards:
        cNo += 1
        numbers = (card.split(":"))[1].split("|")
        
        winNum = [wnum for wnum in numbers[0].split() if wnum]
        myNum = [mnum for mnum in numbers[1].split() if mnum]
        
        numOfMatch = 0
        for i in range(len(winNum)):
            for j in range(len(myNum)):
                if myNum[j] == winNum[i]:
                    numOfMatch += 1
        
        cnt = totalScratchCards[cNo]
        
        while (cnt>0):
            st = cNo+1
            end = st+numOfMatch
            
            for k in range(st, end):
                totalScratchCards[k]+=1
            
            cnt -= 1
        
    for s in range(1, len(cards)+1):
        totalSC += totalScratchCards[s]
    
    print(f"total scratch cards: {totalSC}")

inpfile.close()

File: 2023/cpp/42.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> winningNum, myNum;

void splitfunc(string card)
{
    string part;
    vector<string> cp;

    istringstream iss(card);
    while (iss >> part) cp.emplace_back(part);

    //begin()+2 means we are skipping card and cardno. 
    vector<string>::iterator it;
    for (it = cp.begin()+2; (*it) != "|"; it++)
    {
        winningNum.emplace_back(*it);
    }

    while (it != cp.end())
    {
        myNum.emplace_back(*(it+1));
        it++;
    }
}

int main()
{
    ifstream inpfile;
    inpfile.open("../test-cases/inp4.txt");

    string card;
    vector<string> cards;

    while (getline(inpfile, card)) cards.emplace_back(card);

    unordered_map<int, int> hashmap;
    for (int i = 1; i < cards.size()+1; i++)
    {
        hashmap[i] = 1;
    }

    int cNo, totalScratchCards;
    cNo = totalScratchCards = 0;

    for (auto card : cards)
    {
        cNo++;
        splitfunc(card);

        int numOfMatch = 0;
        for (int i = 0; i < winningNum.size(); i++)
        {
            for (int j = 0; j < myNum.size(); j++)
            {
                if (myNum[j] == winningNum[i]) numOfMatch++;
            }
        }

        int cnt = hashmap[cNo];

        while (cnt > 0)
        {
            int st = cNo+1, end = st+numOfMatch;
            for (int k = st; k < end; k++)
            {
                hashmap[k]++;
            }
            cnt--;
        }
    }

    for (int i = 1; i < cards.size()+1; i++)
    {
        totalScratchCards += hashmap[i];
    }

    cout << "Total Scratch Cards are: " << totalScratchCards;
    cout << endl;
    inpfile.close();
    return 0;
}